### ISSUE TITLE
Dart-like streams use-case

### ISSUE THREAD

**Issue by soywiz** [Kotlin/kotlin-coroutines#40]
I think that the current proposal would allow to easy create/use dart-like streams (except the `await for` sugar). And just wanted to add that possible use-case :) :

https://www.dartlang.org/docs/tutorials/streams/
That are like generators but for asynchronous code.

Example of creating a stream:

```
fun query(query:String) = stream {
    val imperativeCursor = imperativeQuery(query)
    while (await(imperativeCursor.readedAllRowsAsync())) {
        val row = await(imperativeCursor.readDbRowAsync())
        emit(row) // emits row
    }
}
```

Example of using it (`await for` won't work as it does in dart but since functions are that clean we could do something like this):

```
// Reading rows one by one without having to have all in memory (really huge table)
await(query("SELECT * FROM users;").filter { it["name"] != "test" }.mapToClass<User>().each { user ->
    println("${user.name} : ${user.age}")
})
// Sum all ages asynchronously and with a functional style without having to keep all the data in memory
val agesSum = await(query("SELECT age FROM users;").mapToClass<User>().filter { it.name != "test" }.mapReduce(0) { acc, item -> acc + item.age })
println("Sum of all ages: $agesSum")
```

**Comment by soywiz** [comment]
I have published an example of this, here (working with 1.1 from git):
https://github.com/kotlin-es/coroutine-examples/blob/master/src/example-async-generator.kt

**Comment by elizarov** [comment]
Seems to be even better with 1.1-M04, because you can now defined `operator iterators that brings in `for` support that works just like `await for` in Dart as shown here: https://github.com/Kotlin/kotlin-coroutines/blob/revision-2.1/examples/asyncGenerate.kt
